### ISSUE TITLE
Qwen3.5模型训练显存过高问题排查及解决方案

### ISSUE THREAD

**Issue by shen-jiabin** [modelscope/ms-swift#9124]
### Checklist / 检查清单

- [x] I have searched existing issues, and this is a new question or discussion topic. / 我已经搜索过现有的 issues，确认这是一个新的问题与讨论。

### Question Description / 问题描述

说明下我的问题，我是用8卡*80G/卡的硬件，对qwen3.5-4b模型做sft训练，样本长度约为20k，开启了所有降低显存的参数配置后，8卡显存仍然打满，因此做了排查，最终发现原因是liger-kernel对qwen3.5模型并未生效（相关pr已经在liger-kernl的main分支中，但0.7.0版本中仍不包含），并通过简单修改本地代码的方式解决了，供大家参考

[fix_gpu_memory.md](https://github.com/user-attachments/files/26777655/fix_gpu_memory.md)

**Comment by fyw1999** [comment]
请问友友这个会影响GRPO的显存占用嘛

**Comment by shen-jiabin** [comment]
> 请问友友这个会影响GRPO的显存占用嘛

                                                                                                                                                                                                                     
    具体来说：                                                                                                         
                                                                                                                       
    受影响的训练方式：                                                                                                 
    - SFT — 直接计算 CE loss（已验证）                                                                               
    - GRPO/PPO — policy model 的 loss 计算同样经过 lm_head → CE，而且 GRPO 还需要计算 reference model 的 log_prob，也走
     lm_head，显存压力更大                                                                                             
    - DPO/KTO — 需要同时计算 chosen 和 rejected 的 log_prob，两次 lm_head forward                                      
    - Pretrain — 同样走 CE loss                                                                           
                                                                                                                       
    根因是一样的： liger-kernel 0.7.0 未注册 qwen3_5 model_type，导致 --use_liger_kernel true                          
    对所有训练方式都是静默空操作。标准 F.cross_entropy 对 [batch, seq_len, 248320] 的 fp32                             
    膨胀不会因训练方式不同而消失。                                                                                     
                                                                                                                       
    实际上 #8450   就是 Qwen3.5-9B 做 GRPO 训练时遇到的显存异常高的问题，大概率是同一个根因。

**Comment by shen-jiabin** [comment]
---                                                                                                   
                                                                                                        
  ## 后续发现：Megatron 后端训练的两个上游显存泄漏                                                      
                                                                                                        
  上面的 liger-kernel 问题解决后，切换到 Megatron 后端继续训练，发现了两个**独立的上游显存泄漏**，与    
  ms_swift 自身代码无关，但会影响所有使用 Megatron 后端的 ms_swift 用户。                               
                                                                                                        
  ---                                                                                                   
   
  ### 泄漏 1：TransformerEngine CrossEntropyFunction 未调用 `mark_dirty()`                              
                                                            
  **影响条件**：使用 TE 的 cross entropy 实现（`cross_entropy_fusion_impl='te'`）+ 梯度累积（多         
  micro-batch）                                                                                         
   
  **现象**：每个 micro-batch 泄漏一个完整的 logits tensor（数百 MiB），逐 step 累积导致 OOM             
                                                            
  **根因**：TE 的 `CrossEntropyFunction.forward()` 用 triton 内核原地修改输入 logits，但未调用          
  `ctx.mark_dirty(inp)`，导致 PyTorch C++ autograd 引擎永久保留输入引用
                                                                                                        
  **上游 issue**：https://github.com/NVIDIA/TransformerEngine/issues/2899                               
   
  **临时 workaround**：                                                                                 
                                                            
  修改 `transformer_engine/pytorch/cross_entropy.py`，在 forward 中追踪输入，backward 后释放 storage：  
                                                            
  ```python                                                                                             
  import collections                                        
  _leaked_ce_inputs = collections.deque()                                                               
                                                                                                        
  def _drain_leaked_inputs():                                                                           
      while _leaked_ce_inputs:                                                                          
          t = _leaked_ce_inputs.popleft()                                                               
          if t.untyped_storage().size() > 0:                                                            
              t.storage().resize_(0)                                                                    
          t.detach_()                                                                                   
          del t                                                                                         
                                                                                                        
  # CrossEntropyFunction.forward() 中，triton 调用前添加：                                              
  _leaked_ce_inputs.append(inp)                                                                         
  ```                                                                                                   
                                                                                                        
  在 `megatron/core/pipeline_parallel/schedules.py` 每个 `backward_step()` 后调用                       
  `_drain_leaked_inputs()`。                                
                                                                                                        
  ---                                                       

  ### 泄漏 2：`make_viewless_tensor()` + Context Parallel 导致 hidden_states 泄漏                       
   
  **影响条件**：`context_parallel_size > 1`                                                             
                                                                                                        
  **现象**：每个 micro-batch 泄漏一个 hidden_states tensor（`(T, 1, hidden_size)` bf16），~0.5 GiB/step 
                                                                                                        
  **根因**：CP>1 时 hidden_states 是 view（来自序列分片），`make_viewless_tensor()` 对其做 clone
  产生独立 storage。clone 的 tensor 流经 TE 的 `OperationFuser`，被设置 `_do_not_clear = True` 且在     
  backward 后不清除。C++ autograd 引擎保留引用，storage 永久泄漏。                                      
                                                                  
  **上游 issue**：计划提交至 https://github.com/NVIDIA/Megatron-LM/issues                               
                                                                         
  **临时 workaround**：                                                                                 
                       
  修改 `megatron/core/utils.py`：                                                                       
                                                            
  ```python                                                                                             
  import collections as _collections                        
  _leaked_viewless_tensors = _collections.deque()                                                       
                                                                                                        
  def _drain_leaked_viewless_tensors():
      while _leaked_viewless_tensors:                                                                   
          t = _leaked_viewless_tensors.popleft()            
          if t.untyped_storage().size() > 0:                                                            
              t.storage().resize_(0)                                                                    
          t.detach_()                                                                                   
          del t                                                                                         
                                                                                                        
  def make_viewless_tensor(inp, requires_grad, keep_graph):                                             
      if inp._base is None:                                                                             
          return inp                                                                                    
      if keep_graph:                                                                                    
          out = inp.clone().requires_grad_(requires_grad)                                               
          _leaked_viewless_tensors.append(out)                                                          
          return out                                                                                    
      else:                                                                                             
          return _kernel_make_viewless_tensor(inp, requires_grad)                                       
  ```                                                                                                   
                                                                                                        
  在 `megatron/core/pipeline_parallel/schedules.py` 每个 `backward_step()` 后调用
  `_drain_leaked_viewless_tensors()`。                                                                  
                                                                                                        
  ---
                                                                                                        
  ### 验证结果                                              
                                                                                                        
  两个 workaround 同时生效后，`memory_allocated` 从逐步增长 ~0.5 GiB/step 变为完全稳定（40+ step
  验证，18.14 GiB 零增长），训练速度无影响。                                                            
                                                            
  ### 影响范围                                                                                          
                                                                                                        
  这两个泄漏不限于 ms_swift，所有基于 Megatron Core + TransformerEngine 的框架（NeMo、Pai-Megatron-Patch
   等）在相同配置下都会触发。ms_swift 的 Megatron 后端用户如果遇到训练中显存持续增长或                  
  OOM，可以先尝试上述 workaround。                                                    
                                                                                                        
  ---

**Comment by dv0** [comment]
感谢，确实可以解决！
